### PR TITLE
RavenDB-15693 Allow to boost on complex queries

### PR DESCRIPTION
**Pull request by ayende**

### CODE REVIEW

File: test/SlowTests/Issues/RavenDB-15693.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                .AndAlso()
+                .OpenSubclause()
+                .Search(x => x.StrVal2, "b")
+                .OrElse()
+                .Search(x => x.StrVal3, "search")
+                .CloseSubclause()
+                .Boost(0.2m);
+            var queryBoost = q.ToString();
+            
+            Assert.Equal("from 'Docs' where search(StrVal1, $p0) and boost(search(StrVal2, $p1) or search(StrVal3, $p2), $p3)", queryBoost);
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-15693.cs:33]: Query without boost would look like:
```
from 'Docs' where search(StrVal1, $p0) and (search(StrVal2, $p1) or search(StrVal3, $p2))
```

The missing `( )` inside the boost expression will not affect how Lucene generates (and then interprets) query?

Comment (ayende) [reply test/SlowTests/Issues/RavenDB-15693.cs:33]: No, the `()` is there to ensure order of operations, the `boost()` serves the same purpose.

**Comment by ppekrol** [comment]
@ayende can you check failing tests? Also does it affect backward compatibility?

**Comment by ayende** [comment]
Rebased on the latest and re-run those queries, they all work. Very strange. Will test again on CI.

I don't believe that this should have any impact on backward compatability of queries. Such queries were invalid, so cannot happen.

File: src/Raven.Client/Documents/Session/Tokens/CloseSubclauseToken.cs [hunk 2]
```diff
@@ -4,6 +4,7 @@ namespace Raven.Client.Documents.Session.Tokens
 {
     public sealed class CloseSubclauseToken : QueryToken
     {
+        public string BoostParameterName { get; set; }
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/Tokens/CloseSubclauseToken.cs]: Issue with failing tests is here. We are using singleton.

File: src/Raven.Client/Documents/Session/Tokens/OpenSubclauseToken.cs [hunk 3]
```diff
@@ -9,9 +9,12 @@ private OpenSubclauseToken()
         }
 
         public static readonly OpenSubclauseToken Instance = new OpenSubclauseToken();
+        public string BoostParameterName { get; set; }
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/Tokens/OpenSubclauseToken.cs]: Issue with failing tests is here. We are using singleton.

**Comment by ppekrol** [comment]
Fixed.

**Pull request merged**